### DISCUSSION TITLE
UAC2 Asynchronous Mode

### DISCUSSION THREAD

**Discussion by orlyprofili** [hathach/tinyusb#2207] [Q&A]
I have put out products with UAC2 asynchronous mode (audio clocked by device) with XMOS.  Have done UAC1 asynchronous with STM32 the hard way, long before tinyusb or even decent vendor code was available, and with the terrible state of Windows audio at the time.

I spoke with @HiFiPhile a bit on this topic with respect to tinyusb, he was very helpful, pointing out the limitations of some usb phys. It seems like it's a good time to help push this forward, and I'd like to help.

Is there a dev board and known config/example that works, or is known to be the most functional and reliable? If so, I'd like to get it, confirm and test, fix and/or document limitations, etc.  I'd like to be able to see where it maxes out (sample rate, I/O count, bit depth), how it plays when part of a composite device, so on.  I'd like to focus on only supporting mcu's with phys that are really suitable for this purpose, as some usb phys are just not worth chasing.  I'm guessing some NXP (LPC or RT series).

As it is right now, it seems that xmos is still the leading and maybe only truly reliable and robust solution for uac2 asynch mode, and tinyusb (maybe zephyr) has the potential to change that.

**Comment by HiFiPhile** [comment]
> UAC2 asynchronous mode (audio clocked by device) 

There is no additional difficulty using asynchronous mode under TUSB, only need to calculate the feedback value correctly :)


> how it plays when part of a composite device

In my personal Muse-1 project :
- 1nd UAC instance 32/384k out
- 2nd UAC instance 24/48k mic in
- DFU runtime
- Debug interface for status display (see below)
```
E:\mcu\lpc55\project\Muse-1_FW>debug.py sl
DC_Offset
    -10.60 -5.16 -5.49 -6.84
Sample_Rate
    48000
Bytes_Sample
    4
Resolution
    32
Volume
    -9.5dB
DoP_Mode
    False
Playback
    True
Record
    False
Volume_Mic
    30.1dB
Buffer_Count
    467
```

> Is there a dev board and known config/example that works, or is known to be the most functional and reliable?

I believe the best board to start with is LPCXpresso55S69, LPC55 has a performant USB phy and good peripherals such as multi-channel I2S capable more than 4 data lines. In my application it can easily handle 32/384k streaming at about 15% CPU usage while the 2nd core used as SPDIF software encoder.

Although i.MXRT1064 could be the ultimate choice it's too overkill and expensive unless complexe EQ is needed.

However all generic dev boards are not made for high-res audio application, some boards have a basic codec IC but can't handle high sample rate. To make a neat solution I think we to make a "reference board" with a good clock generator, multi-channel I2S breakout and jumper selectable ADC/DAC/Mic.

Something like this:
![uac drawio (1)](https://github.com/hathach/tinyusb/assets/4375114/20560c43-ea33-4a64-b4e4-6822b9ca75c1)

**Reply by orlyprofili** [reply]
Interesting. Maybe an audio adc+dac+local clock board that can easily wired into the lpc55 eval board?

**Comment by HiFiPhile** [comment]
> Maybe an audio adc+dac+local clock board that can easily wired into the lpc55 eval board?

I just looked at the board schematic, it's indeed possible to have:
- 1 Set of 2pin(4ch) I2S Tx
- 1 Set of 1pin(2ch) I2S Rx
- 1 SPDIF Tx with software encoder

If I followed the pinout right...

**Comment by orlyprofili** [comment]
yes, and can remove a 0R resistor to replace the MCLK with something higher
quality as well.  I will look into doing this soon.

**Comment by HiFiPhile** [comment]
> yes, and can remove a 0R resistor to replace the MCLK with something higher

Or just use P19 pin4.

Here is a possible pinout:
![image](https://github.com/hathach/tinyusb/assets/4375114/6becb463-fbf6-411d-b777-627fe2f4d4bd)

MCUXpresso Config Tool project:
[LPC55S69_pin.zip](https://github.com/hathach/tinyusb/files/12285713/LPC55S69_pin.zip)

**Comment by hathach** [comment]
just want to chime in, although I am not really familliar with audio, but I will try to help whenever possible. Maybe it is worth to add bsp i2s, pdm API to get a pratical examples with some supported boards. Audio is proabably is too hard for un-familliar to figure out by their own (including me).

**Comment by HiFiPhile** [comment]
> Maybe it is worth to add bsp i2s, pdm API to get a pratical examples with some supported boards.

Yes it would be helpful.

One difficulty is dev boards often use a highly integrated codec ic which is not trivial to configurate, the resulting example will be more like "how to use codec X". (I often unpopulate them and test I2S signal)

Or we can get basic I2S working and ignore on-board codec.

**Reply by orlyprofili** [reply]
I just realized this board is probably the most usable in it's existing form, it has a better codec and 6in/8out to cover a broader range of use cases, and to show how good uac2 asynch on tinyusb can be.  The chip has a dedicated audio subsystem PLL which makes it easy to switch between something decent (generated by mcu), or if you want hifi excellence (external precision audio clock).

https://www.nxp.com/design/development-boards/i-mx-evaluation-and-development-boards/i-mx-rt600-audio-evaluation-kit:MIMXRT685-AUD-EVK

what do you think @HiFiPhile ?

**Reply by HiFiPhile** [reply]
Didn't know this board exist :)

While RT600 is very powerful and this board is good at multi-channel, this SoC is targeting more at professionals than DIYers.
- RT600's package is very tiny, even VFBGA176 is too hard for low cost PCB fabs like JLC
- It doesn't have internal flash and need a PMIC for complexe power rails, make DIY projects more difficult

I've 10 years+ PCB design experience for SoC I still need to be very careful.

Personally I'd like to make a demo allowing users to focus on uac2 itself as other parts of the systems are obvious. LPC55 is a good compromise between performance and easy to use, one 3.3V power supply is enough and LQFP package is hand solderable (while I'm using BGA for space-saving).

For clocking now I'm thinking more about GPIO controlled 45.1584M/49.152M oscillators rather than clock generator IC I posted above.

Dumb DAC like PCM5102A is very beginner friendly, I2S MICs like ICS43434 serve the role of dumb ADC without external input needed. While user can bypass them with jumper.

**Comment by orlyprofili** [comment]
I was thinking the exact same for clock (gpio mux) and wrote that earlier, but didn’t send, as I wanted to think on it more :) I hear your points regarding future design beyond eval kit and to a personal custom project. But…For the benefit of the tinyusb project, the rt600 dev kit is more complete, allowing easier on going development, whereas the lpc55 version we’ve been discussing would require also taking design files to some fab/assembly house (eg jlcpcb) for this custom add-on board we’ve been talking about. I’m not adamant about it, and could go either way, ultimately just interested in hearing and testing a working tinyusb uac2 in action. On Aug 11, 2023, at 2:41 PM, HiFiPhile ***@***.***> wrote:﻿
Didn't know this board exist :)
While RT600 is very powerful and this board is good at multi-channel, this SoC is targeting more at professionals than DIYers.

RT600's package is very tiny, even VFBGA176 is too hard for low cost PCB fabs like JLC
It doesn't have internal flash and need a PMIC for complexe power rails, make DIY projects more difficult

I've 10 years+ PCB design experience for SoC I still need to be very careful.
Personally I'd like to make a demo allowing users to focus on uac2 itself as other parts of the systems are obvious. LPC55 is a good compromise between performance and easy to use, one 3.3V power supply is enough and LQFP package is hand solderable (while I'm using BGA for space-saving).
For clocking now I'm thinking more about GPIO controlled 45.1584M/49.152M oscillators rather than clock generator IC I posted above.
Dumb DAC like PCM5102A is very beginner friendly, I2S MICs like ICS43434 serve the role of dumb ADC without external input needed. While user can bypass them with jumper.

**Comment by HiFiPhile** [comment]
>  rt600 dev kit is more complete, allowing easier on going development

Of cause it's an option.  

While RT600 isn't supported by TUSB yet, it looks more like an overclocked version of lpc55 than rt10xx as they have the same core, usb phy and most peripherals.

Adding TUSB support should be easy (maybe it _could_ work using lpc55 compile option) but I'm not interested in spending $170 for the board just to make uac2 examples.  Of cause I'll help you to get it work :)

**Comment by orlyprofili** [comment]
Thanks @HiFiPhile, I'll be in touch :)
I can also make time to make a draft for the extension board for lpc55, will share while in progress
